---
title: "Differential Gene Expression Analysis -RNA_seq_DEA-"
author: Viivi Castrén
output: html_document
date: "2024-05-21"
---

# 1. Set-up 

#Install R and RStudio and open RStudio
#Go to File and create New project:
#--> create New Directory --> New Project --> name it and “Create the project as subdirectory of:” the Desktop

#Check the current R version. with "R.Version()"

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
#Check in which directory you are:
```{r}
setwd("C:/Users/Käyttäjä/Desktop/RNA_seq_DEA") 
```
# 2. Count normalization
## Installation of packages:
```{r, eval=FALSE}
if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("DESeq2")

if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("edgeR")
```
## Reading data files:
#Read raw data: (modify the code to correct location of the data in my computer)

```{r}
Coinfection.targets<-read.delim("./data/fileDesc.txt")

rownames(Coinfection.targets)<-c("Ha1","Ha2","Ha3","Ctr1","Ctr2","Ctr3")

```
# Loading EdgeR
```{r}
library(edgeR)

```
#Reading .txt files:
```{r}
Coinfection.orig <- readDGE(Coinfection.targets, header=F)
```

#Dimension of the data set:
```{r}
dim(Coinfection.orig)
head(Coinfection.orig)
```

#Checking first 6 rows of the data:
```{r}
Coinfection.rawCount <- Coinfection.orig$count
dim(Coinfection.rawCount)
```

```{r}
head(Coinfection.rawCount)
```

##Building meta data:

#We define sampletype: we only have two types of samples, Control and Ha-infection

```{r}
sampletype <- factor(c(rep("Ha",3), rep("Ctr", 3)))
```
# Build meta data frame:
```{r}
meta <- data.frame(sampletype, row.names = colnames(Coinfection.orig$count))
```
#Column name of counts dataframe:
```{r}
colnames(Coinfection.orig$count)
```
#Rowbanes:
```{r}
rownames(meta)
```
#Checking if sample names match in both files:
```{r}
all(colnames(Coinfection.orig$count) %in% rownames(meta))
```
# Creating DESeqDataset object:
```{r, message=FALSE}
library(DESeq2)
```


```{r}
dds <- DESeqDataSetFromMatrix(Coinfection.orig, colData = meta, design = ~ sampletype)
```

```{r}
head(counts(dds))
```


## Generate size factor:
#The median of ratios method of normalization (DESeq2 has a single estimateSizeFactors() function that will generate size factors):
```{r}
dds <- estimateSizeFactors(dds)
sizeFactors(dds)
```

## Generate the normalized counts:
#Retrieve the normalized counts matrix from dds, we use the counts() function and add the argument normalized=TRUE:
```{r}
normalized_counts <- counts(dds, normalized=TRUE)
```
#Save the normalized_countsinto results to your local path:
```{r}
write.csv(normalized_counts, file="./results/coinfection_normalized_counts_DESeq2.csv")
```
# 3. Sample-level quality control:
#Dataset used for PCA:
#Transform normalized counts for the dataset:
```{r}
rld <- rlog(dds, blind=TRUE)
plotPCA(rld, intgroup="sampletype")
```
#Saving the plot as a PDF file:
```{r}
pdf("./results/PlotPCA_dds.pdf")
plotPCA(rld, intgroup="sampletype")
dev.off()
```
## Hierarchial Clustreing heatmap:
#Extract the rlog matrix from the object using assay function
```{r}
rld_mat <- assay(rld)
```
#Compute the pairwise correlation values for all the samples:
```{r}
rld_cor <- cor(rld_mat) 
```
#Checking the output:
```{r}
head(rld_cor)
head(meta)
```

#Install pheatmap package and load the package
```{r, eval=FALSE}
install.packages("pheatmap")
```

#Plot heatmap using the correlation matrix and the metadata object:
```{r}
library(pheatmap)
pheatmap(rld_cor, annotation = meta)
```

#Changing the colors:
```{r}
heat.colors <- RColorBrewer::brewer.pal(6, "Blues")
pheatmap(rld_cor, annotation = meta, color = heat.colors, border_color=NA, fontsize = 10, 
         fontsize_row = 10, height=20)
```

#Saving the heatmap in my computer into results folder:
```{r}
pdf("./results/PlotHeatmap_dds.pdf")
heat.colors <- RColorBrewer::brewer.pal(6, "Blues")
pheatmap(rld_cor, annotation = meta, color = heat.colors, border_color=NA, fontsize = 10, 
         fontsize_row = 10, height=20)
```

# 4. Differential expression analysis (DEA) using EdgeR:
#What is the current directory?
```{r}
getwd()
```

#Loading EdgeR package:
```{r}
library(edgeR)
options(digits=3)
```

#To tell R where the data files are:
```{r}
infection.targets<-read.delim("./data/fileDesc.txt")
```

#Checking infection targets:
```{r}
infection.targets
```

#Changing raw names of the dataframe and checking them:
```{r}
rownames(infection.targets)<-c("Ha1","Ha2","Ha3","Ctr1","Ctr2","Ctr3")
infection.targets
```

#Read and merges a set of text files containing gene expression counts, which makes a DGEList object directly:
```{r}
infection <- readDGE(infection.targets, header=F)
```

#Check the dimension of DGElist R object:
```{r}
dim(infection)
head(infection)
```
#Get the raw mapped count before filtering
```{r}
infection.rawCount <- infection$count
head(infection.rawCount)
```

#Install ggplot2 package:
```{r, eval=FALSE}
install.packages("ggplot2")
```

```{r}
library(ggplot2)
```

#How RNA-seq counts are distributed, plot a histogram of the counts for a single sample, ‘Ha1’:
```{r}
ggplot(infection.rawCount) +
  geom_histogram(aes(x = Ha1), stat = "bin", bins = 200) +
  xlab("Raw expression counts") +
  ylab("Number of genes")
```

#Export the .png file

```{r}
png("./results/count distribution.png", res=300, height=1800, width=1800)
ggplot(infection.rawCount) +
  geom_histogram(aes(x = Ha1), stat = "bin", bins = 200) +
  xlab("Raw expression counts") +
  ylab("Number of genes")
dev.off()
```

#Export raw count table into results folder
```{r}
write.csv(infection.rawCount, file="./results/infection.rawCounts.csv")
```

#The counts per million (TMM normalised) before filtering:
```{r}
infection.normCPM <- cpm(calcNormFactors(infection))
dim(infection.normCPM)

head(infection.normCPM)
write.csv(infection.normCPM, file="./results/infection.normCPM.csv")
```

#Keep genes that are expressed at least 1 CPM in at least 3 libraries, normally it is the number of biological replicates of smaller group:
```{r}
infection.filtered <- rowSums(cpm(infection)>1) >=3
table(infection.filtered)
```
#Libraries size of data BEFORE filtering
```{r}
infection$samples$lib.size
```

#Cover the original file with our filter data, and get the size of data after filtering;
```{r}
Infection <- infection[infection.filtered,]
colSums(Infection$counts)
dim(Infection)
```

#Update the filtered libraries size
```{r}
Infection$samples$lib.size <- colSums(Infection$counts)
Infection$samples
```
#Normalisation with TMM method, and librarier after normalisation:
```{r}
Infection = calcNormFactors(Infection)
Infection$samples
```
#Counts per million (TMM normalised) after filtering
```{r}
Infection.filtered.normCPM <-cpm(calcNormFactors(Infection))
```
#Export TMM normalized count table after filtering
```{r}
write.csv(Infection.filtered.normCPM, file="./results/Infection.filtered.normCPM.csv")
```


## Differentailly gene expression analysis:

#Treatment factor:
```{r}
group<-factor(c('Ha','Ha','Ha',"Ctr","Ctr","Ctr"))
```
#Describing the experimental design,one factor with intercept, here Ctr is the intercept
```{r}
Infection.design <- model.matrix(~group)   
rownames(Infection.design)<-colnames(Infection$counts)
Infection.design
```
#Checking if some samples are outliner. An MDS plot shows the relative similarities of the six samples.
```{r}
plotMDS(Infection, main="MDS plot of RNA-Seq", labels=colnames(Infection$counts))
```
#Export the MDS plot to a PNG file
```{r}
png("./results/plotMDS.Infection.png", res=300, height=1800, width=1800)
plotMDS(Infection, main="MDS plot of Infection RNA-Seq", labels=colnames(Infection$counts))
dev.off()
```
#Estimating the common dispersion, trended dispersion, tagwwise dispersion Dispersion means biological coeffient of variation (BCV) squared:
```{r}
Infection <- estimateGLMCommonDisp(Infection, Infection.design)
Infection <- estimateGLMTrendedDisp(Infection, Infection.design)
Infection <- estimateGLMTagwiseDisp(Infection, Infection.design)
```
#Creating a visual representation of the mean-variance relationship and BCV-mean logCPM relationship
```{r}
plotMeanVar(Infection, show.tagwise.vars=T,NBline=T)
plotBCV(Infection)
```
#Fit DGEList and design matrix to genewise GLM
```{r}
Infection.fit <- glmFit(Infection, Infection.design)
colnames(Infection.fit)
```
#Likelihood test for genes respond differently between different conditions, table of top differentially expressed tags, n specify n tags to display
```{r}
lrt.Ha_vs_Ctr <- glmLRT(Infection.fit, coef=2)  
```
#To compare Ha vs Ctr (Ha_vs_Ctr)
```{r}
t1<-topTags(lrt.Ha_vs_Ctr, n=nrow(Infection))
head(t1$table)
```
#Total number of differentially expressed (DE) genes in different comparsion at FDR < 0.05
```{r}
summary(decideTests(lrt.Ha_vs_Ctr, adjust.method="BH", p.value=0.05))
```
#The UP-regulated genes (can change logFC to be more strict)
```{r}
nrow(subset(topTags(lrt.Ha_vs_Ctr, n=586)$table,  logFC > 0))
lrt.Ha_vs_Ctr_UP <- subset(topTags(lrt.Ha_vs_Ctr, n=586)$table, logFC > 0)
```
#The DW-regulated genes (can change logFC to be more strict)
```{r}
nrow(subset(topTags(lrt.Ha_vs_Ctr, n=586)$table,  logFC < 0))
lrt.Ha_vs_Ctr_DW <- subset(topTags(lrt.Ha_vs_Ctr, n=586)$table, logFC < 0)
```
#Differentially expressed transcripts’ tag
```{r}
DEtags.lrt.Ha_vs_Ctr <- rownames(Infection)[as.logical(decideTests(lrt.Ha_vs_Ctr, adjust.method="BH", p.value=0.05))]
```
#Export list of UP-regulated and DW-regulated transcripts
```{r}
write.csv(lrt.Ha_vs_Ctr_UP, file="./results/lrt.Ha_vs_Ctr_UP.csv")
write.csv(lrt.Ha_vs_Ctr_DW, file="./results/lrt.Ha_vs_Ctr_DW.csv")
```
#To ask all the genes label as grey color
```{r}
Infection.colHavsCtr = rep('grey55', nrow(Infection))
```
#To assign colour for DE transcripts
```{r}
Infection.colHavsCtr[lrt.Ha_vs_Ctr$table$PValue < 0.05 & lrt.Ha_vs_Ctr$table$logFC >0 ] <- "red"
Infection.colHavsCtr[lrt.Ha_vs_Ctr$table$PValue < 0.05 & lrt.Ha_vs_Ctr$table$logFC <0 ] <- "blue"
```
#Plot Smear plot with adjustment on Y-axis label
```{r}
par(omi=c(0.1,0.1,0.1,0.1), las=1, cex=0.5, mgp=c(3,1,0), cex.main=1.8, cex.lab=1.4, cex.axis=1.4)
plotSmear(lrt.Ha_vs_Ctr, de.tags=DEtags.lrt.Ha_vs_Ctr, xlab="log-counts per million (logCPM)", ylab="log2-fold change (log2FC)", main="Ha infection compared to Control", pch=19, cex=0.4, smearWidth=0.5, panel.first=grid(), smooth.scatter=FALSE, ylim=c(-7,7), yaxs="i")

abline(h=c(-1,1),col="dodgerblue")
```
#Plot Smear plot with adjustment on Y-axis label and colour of DE tags changed
```{r}
par(omi=c(0.1,0.1,0.1,0.1), las=1, cex=0.5, mgp=c(3,1,0), cex.main=1.8, cex.lab=1.4, cex.axis=1.4)
plotSmear(lrt.Ha_vs_Ctr, xlab="log-counts per million (logCPM)", ylab="log2-fold change (log2FC)", main="a infection compared to Control", smearWidth=0.5, pch=21, cex=0.4, deCol="red", col=Infection.colHavsCtr, ylim=c(-7,7), yaxs="i")

abline(h=c(-1,1),col="dodgerblue")
```
#Plot Smear plot with adjustment on Y-axis label and colour of DE tags changed and export as png:

```{r}
png("./results/plotSmear.InfectionRNAseq.png", res=300, height=1800, width=1800)
par(omi=c(0.1,0.1,0.1,0.1), las=1, cex=0.5, mgp=c(3,1,0), cex.main=1.8, cex.lab=1.4, cex.axis=1.4)
plotSmear(lrt.Ha_vs_Ctr, xlab="log-counts per million (logCPM)", ylab="log2-fold change (log2FC)", main="Ha infection compared to Control", smearWidth=0.5, pch=21, cex=0.4, deCol="red", col=Infection.colHavsCtr, ylim=c(-7,7), yaxs="i")
abline(h=c(-1,1),col="dodgerblue")
dev.off()
```

